pkg/query-service/app/dashboards/model.go

```diff
@@ -233,10 +267,58 @@ func UpdateDashboard(uuid string, data map[string]interface{}) (*Dashboard, *mod
 		zap.S().Errorf("Error in inserting dashboard data: ", data, err)
 		return nil, &model.ApiError{Typ: model.ErrorExec, Err: err}
 	}
-
+	if existingCount != newCount {
+		// if the count of trace and logs panel has changed, we need to update feature flag count as well
+		updateFeatureUsage(fm, newCount-existingCount)
+	}
 	return dashboard, nil
 }
 
+func updateFeatureUsage(fm interfaces.FeatureLookup, usage int64) *model.ApiError {
+	feature, err := fm.GetFeatureFlag(model.QueryBuilderPanels)
+	if err != nil {
+		switch err.(type) {
+		case model.ErrFeatureUnavailable:
+			zap.S().Errorf("feature unavailable", zap.String("featureKey", model.QueryBuilderPanels), zap.Error(err))
+			return model.BadRequest(err)
+		default:
+			zap.S().Errorf("feature check failed", zap.String("featureKey", model.QueryBuilderPanels), zap.Error(err))
+			return model.BadRequest(err)
+		}
+	}
+	feature.Usage += usage
+	if feature.Usage == feature.UsageLimit {
+		feature.Active = false
+	}
```

Comment:
This should be `>=` instead. One could send a request with number panels that make the Usage go higher that limit.